docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,251 @@
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `RequestPrepareProposal`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `RequestPrepareProposal`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`RequestPrepareProposal.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. However, how an application determines
+validity of a proposed block depends on the application and its varying use cases.
+For most applications, simply calling the `AnteHandler` chain would suffice, but
+there could easily be other applications that need more control over the validation
+process of the proposed block, such as ensuring txs are in a certain order or
+that certain transactions are included. While this theoretically could be achieved
+with a custom `AnteHandler` implementation, it's not the cleanest UX or the most
+efficient solution.
+
+Instead, we will define an additional ABCI interface method on the existing
+`Application` interface, similar to the existing ABCI methods such as `BeginBlock`
+or `EndBlock`. This new interface method will be defined as follows:
+
+```go
+ProcessProposal(sdk.Context, abci.RequestProcessProposal) error {}
+```
+
+Note, we must call `ProcessProposal` with a new internal branched state on the
+`Context` argument as we cannot simply just use the existing `checkState` because
+`BaseApp` already has a modified `checkState` at this point. So when executing
+`ProcessProposal`, we create a similar branched state, `processProposalState`,
+off of `deliverState`. Note, the `processProposalState` is never committed and
+is completely discarded after `ProcessProposal` finishes execution.
+
+We will only populate the `Status` field of the `ResponseProcessProposal` with
+`ACCEPT` if ALL the transactions were accepted as valid, otherwise we will
+populate with `REJECT`.
+
+### `DeliverTx`
+
+Since transactions are not truly removed from the app-side mempool during
+`PrepareProposal`, since `ProcessProposal` can fail or take multiple rounds and
+we do not want to lose transactions, we need to finally remove the transaction
```

Comment:
How exactly will transactions be removed from the app-side mempool during `DeliverTx`? Also, what about removing transactions during `RecheckTx`? These are the two code paths which commonly remove transactions from the Mempool. How would this be handled in the app-side mempool?

I wonder if we should add another method to the `Mempool` interface to support this like this following? 

```go
type Mempool[T MempoolTx] interface {
	// Remove attempts to remove a MempoolTx from the app-side mempool returning
	// an error upon failure. It called when a transaction is committed in a block via `DeliverTx`,
    // or when a transaction returns a non-zero Code for `RecheckTx`.
	Remove(T) error
```

Perhaps Cosmos could call this function under the hood in two cases:
1. The transaction is included in a block (`DeliverTx`).
2. `RecheckTx` fails for the transaction (returns non-zero).

> Alternatively, we can keep the transactions as truly being removed during the
reaping phase in `PrepareProposal` and add them back to the app-side mempool in
case `ProcessProposal` fails.

This is not a great solution as `PrepareProposal` is only called for the current proposer. Other non-validator nodes like full-nodes would have no way to remove transactions from their app-side mempool.

Reply:
> How exactly will transactions be removed from the app-side mempool during DeliverTx? 

For each tx, find it in the mempool and remove it. Yeah, seems like we need to add a `Remove` method to the interface.

> Also, what about removing transactions during RecheckTx?

You can't. `recheck` is a local operator configuration and is not always called.

>  Other non-validator nodes like full-nodes would have no way to remove transactions from their app-side mempool.

They would as they all call `DeliverTx`, which is where they would remove the tx (if it exists).

Reply:
> They would as they all call DeliverTx, which is where they would remove the tx (if it exists).

This would not be ideal for us. We use the mempool for transactions that are never intended to be included in any block. If we're relying only on `DeliverTx`, then this means we'd have no way to eject those transactions from the app-side mempool.

> You can't. recheck is a local operator configuration and is not always called.

That's true. It is a local operator configuration, but Cosmos knows if a transaction is being rechecked via the [runTxMode](https://github.com/cosmos/cosmos-sdk/blob/ad9e5620fb3445c716e9de45cfcdb56e8f1745bf/baseapp/baseapp.go#L555-L555). So it seems feasible to do a call to `Remove` [here](https://github.com/cosmos/cosmos-sdk/blob/ad9e5620fb3445c716e9de45cfcdb56e8f1745bf/baseapp/baseapp.go#L756) if the `handler` returns an error, and `ctx.isRecheckTx` is `true`. What do you think?

It might also be convenient to pass a flag to `Remove` indicating for which reason the transaction was removed (`DeliverTx` or `RecheckTx`).

Reply:
> That's true. It is a local operator configuration, but Cosmos knows if a transaction is being rechecked via the [runTxMode](https://github.com/cosmos/cosmos-sdk/blob/ad9e5620fb3445c716e9de45cfcdb56e8f1745bf/baseapp/baseapp.go#L555-L555). So it seems feasible to do a call to Remove [here](https://github.com/cosmos/cosmos-sdk/blob/ad9e5620fb3445c716e9de45cfcdb56e8f1745bf/baseapp/baseapp.go#L757) if the handler returns an error, and ctx.isRecheckTx is true. What do you think?

I suppose we can call `Remove` too when `CheckTx` fails and `isRecheckTx` is `true`. However, we'd still be calling `Remove` during `DeliverTx`, which would still yield the same problem for you. Is this why you want the flag?

> It might also be convenient to pass a flag to Remove indicating for which reason the transaction was removed (DeliverTx or RecheckTx).

Why is this useful? In any case, we can just provide `Context` to all these mempool methods so no need for hacky APIs.

Reply:
> I suppose we can call Remove too when CheckTx fails and isRecheckTx is true. However, we'd still be calling Remove during DeliverTx, which would still yield the same problem for you. Is this why you want the flag?

👍 This should solve our problem.

It's ok that we're still calling `Remove` during `DeliverTx`. The problem I'm describing occurs for transactions that are _never_ included in any block. That is – we intend to modify our custom in-app mempool implementation to never return these transaction in `ReapMaxBytes`.

> Why is this useful? In any case, we can just provide Context to all these mempool methods so no need for hacky APIs.

Ah true. The `Context` is even better. I don't have any specific use-case in mind for the flag. I don't think we need it. It just seems potentially relevant for a custom mempool implementation to understand why it is being asked to remove a transaction from it's mempool.

Reply:
I've updated the interface to have the mempools take a `Context` should should give you what you need 👍